src/filters/authority.rs

```diff
@@ -0,0 +1,92 @@
+//! Authority (aka `Host` header) filter
+//!
+use crate::filter::{filter_fn_one, Filter, One};
+use crate::reject::{self, Rejection};
+use futures::future;
+pub use http::uri::Authority;
+use std::str::FromStr;
+
+/// Creates a `Filter` that requires a specific authority (target server's
+/// host and port) in the request.
+///
+/// # Example
+///
+/// ```
+/// use warp::{Filter, authority::Authority};
+///
+/// let multihost =
+///     warp::authority("foo.com").map(|| "you've reached foo.com")
+///     .or(warp::authority("bar.com").map(|| "you've reached bar.com"));
+/// ```
+pub fn authority(expected: &str) -> impl Filter<Extract = (), Error = Rejection> + Clone {
+    let expected = Authority::from_str(expected).expect("invalid authority");
+    optional()
+        .and_then(move |option: Option<Authority>| match option {
+            Some(authority) if authority == expected => future::ok(()),
+            _ => future::err(reject::not_found()),
+        })
+        .untuple_one()
+}
+
+/// Creates a `Filter` that looks for an authority (target server's host
+/// and port) in the request.
+///
+/// If found, extracts the `Authority`, otherwise continues the request,
+/// extracting `None`.
+///
+/// Rejects with `400 Bad Request` if the `Host` header is malformed or if there
+/// is a mismatch between the `Host` header and the target URI.
+///
+/// # Example
+///
+/// ```
+/// use warp::{Filter, authority::Authority};
+///
+/// let authority = warp::authority::optional()
+///     .map(|authority: Option<Authority>| {
+///         if let Some(a) = authority {
+///             format!("{} is currently not at home", a.host())
+///         } else {
+///             "please state who you're trying to reach".to_owned()
+///         }
+///     });
+/// ```
+pub fn optional() -> impl Filter<Extract = One<Option<Authority>>, Error = Rejection> + Copy {
+    filter_fn_one(move |route| {
+        // The authority can be sent by clients in various ways:
+        //
+        //  1) in the "target URI"
+        //    a) serialized in the start line (HTTP/1.1 proxy requests)
+        //    b) serialized in `:authority` pseudo-header (HTTP/2 generated - "SHOULD")
+        //  2) in the `Host` header (HTTP/1.1 origin requests, HTTP/2 converted)
+        //
+        // Hyper transparently handles 1a/1b, but not 2, so we must look at both.
+
+        let from_uri = route.uri().authority();
+
+        let name = "host";
+        let from_header = route.headers()
+            .get(name)
+            .map(|value|
+                // Header present, parse it
+                value.to_str().map_err(|_| reject::invalid_header(name))
+                    .and_then(|value| Authority::from_str(value).map_err(|_| reject::invalid_header(name)))
+            );
+
+        future::ready(match (from_uri, from_header) {
```

Comment:
Since checking the headers requires a hashmap lookup, I'd check `from_uri` first, and only if `None`, *then* try to get the value from the headers.

Reply:
I kinda like some strictness and see it fail in case there is a discrepancy between URL and Host, rather than just use one of them.

I would also imagine that in the overall inefficient design of warp where you need to do `::exact` multiple times in subtrees (as opposed to `::optional` once and route to a sub-filter dynamically based on that), a hash lookup is a comparatively tiny inefficiency.

So I would leave it as it is, unless you insist.

Reply:
What you say makes sense, I didn't notice the subtlety in the implementation.